controller/monitor/snapshot_monitor.go

```diff
@@ -0,0 +1,467 @@
+
+		nodeName:      nodeName,
+		eventRecorder: eventRecorder,
+
+		snapshotChangeEventChan: snapshotChangeEventChan,
+		snapshotVerifyTaskChan:  make(chan *snapshotVerifyTask, 1),
+
+		inProgressSnapshotVerifyTasks: map[string]struct{}{},
+
+		settings: setting,
+
+		workers: snapshotVerifyWorkerMax,
+
+		syncCallback:     syncCallback,
+		proxyConnCounter: util.NewAtomicCounter(),
+	}
+
+	go m.Start()
+
+	return m, nil
+}
+
+func (m *SnapshotMonitor) Start() {
+	for i := 0; i < m.workers; i++ {
+		go m.snapshotVerifyWorker()
+	}
+
+	go m.processSnapshotChangeEvent()
+	go m.processPeriodicSnapshotVerify()
+}
+
+func (m *SnapshotMonitor) processSnapshotChangeEvent() {
+	for {
+		select {
+		case <-m.ctx.Done():
+			m.logger.WithField("monitor", monitorName).Info("Closing periodic snapshot verification process")
+			return
+		default:
+			event := <-m.snapshotChangeEventChan
+
+			if event.SnapshotName == "" {
+				engine, err := m.ds.GetVolumeCurrentEngine(event.VolumeName)
+				if err != nil {
+					m.logger.WithField("monitor", monitorName).WithError(err).Errorf("failed to find engine for volume %v", event.VolumeName)
+					continue
+				}
+				m.populateEngineSnapshots(engine)
+			} else {
+				m.snapshotVerifyTaskChan <- &snapshotVerifyTask{
+					volumeName:   event.VolumeName,
+					snapshotName: event.SnapshotName,
+				}
+			}
+		}
+	}
+}
+
+func (m *SnapshotMonitor) processPeriodicSnapshotVerify() {
+	ticker := time.NewTicker(snapshotVerifyProcessPeriod)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-m.ctx.Done():
+			m.logger.WithField("monitor", monitorName).Info("Closing periodic snapshot verification process")
+			return
+		case <-ticker.C:
+			v := m.settings[string(types.SettingNameSnapshotDataIntegrity)]
+			dataIntegrity := v.(string)
+
+			v = m.settings[string(types.SettingNameSnapshotDataIntegrityInterval)]
+			dataIntegrityInterval := v.(int64)
+
+			if dataIntegrity == string(types.SnapshotDataIntegrityDisabled) ||
+				!time.Now().After(m.lastSnapshotVerifiedAt.Time.Add(time.Duration(dataIntegrityInterval)*time.Minute)) {
+				continue
+			}
+
+			m.verifySnapshots()
+
+			m.lastSnapshotVerifiedAt = metav1.Time{Time: time.Now().UTC()}
+		}
+	}
+}
+
+func (m *SnapshotMonitor) verifySnapshots() {
+	m.logger.WithField("monitor", monitorName).Info("Starting verifying snapshots")
+
+	engines, err := m.ds.ListEnginesByNodeRO(m.nodeName)
+	if err != nil {
+		m.logger.WithField("monitor", monitorName).WithError(err).Errorf("failed to list engines on node %v", m.nodeName)
+		return
+	}
+
+	for _, engine := range engines {
+		m.logger.WithField("monitor", monitorName).Infof("Populating engine %v snapshots", engine.Name)
+		m.populateEngineSnapshots(engine)
+	}
+}
+
+func (m *SnapshotMonitor) populateEngineSnapshots(engine *longhorn.Engine) {
+	snapshots := engine.Status.Snapshots
+	for _, snapshot := range snapshots {
+		if snapshot.Name == ediskutil.VolumeHeadName || !snapshot.UserCreated || snapshot.Removed {
```

Comment:
NIT: add a comment that explains why we skip these snapshots. (The snapshot prune issue)

Reply:
Added, thank you.